src/olympia/amo/models.py

```diff
@@ -505,3 +507,50 @@ class LongNameIndex(models.Index):
     """Django's Index, but with a longer allowed name since we don't care about
     compatibility with Oracle."""
     max_name_length = 64  # Django default is 30, but MySQL can go up to 64.
+
+
+class FilterableManyToManyDescriptor(ManyToManyDescriptor):
+
+    def __init__(self, *args, **kwargs):
+        self.q_filter = kwargs.pop('q_filter', None)
+        super().__init__(*args, **kwargs)
+
+    @classmethod
+    def _get_manager_with_default_filtering(cls, manager, q_filter):
+        class ManagerWithFiltering(manager):
+            def get_queryset(self):
+                qs = super().get_queryset()
+                return qs.filter(q_filter or models.Q())
```

Comment:
it's this extra `.filter` here that's causing the extra 3 queries in the failing test.

Even a no-op extra `Q()` triggers the extra 3 queries, semi-confirming it's something query cache related.
```
def get_queryset(self):
    qs = super().get_queryset()
    return qs.filter(models.Q())
```

Reply:
but just `return super().get_queryset()` _doesn't_ trigger any extra queries (so it's not the extra classes or the implementation in general)